bin/seedUserStreaks.ts

```diff
@@ -0,0 +1,177 @@
+  con: DataSource,
+  insertQueue: fastq.queueAsPromised<Partial<User>[], void>,
+  userPointer?: UserPointer,
+  limit: number = BATCH_SIZE,
+): Promise<UserPointer> => {
+  const qb = con
+    .getRepository(User)
+    .createQueryBuilder('u')
+    // fetching createdAt as a string, because Date object in node is milisecond precision, but postgres is microsecond
+    .select(['id', 'timezone', '"createdAt"::text'])
+    .where(
+      userPointer?.createdAt
+        ? {
+            createdAt: MoreThan(userPointer?.createdAt),
+          }
+        : 'TRUE',
+    )
+    .orWhere(
+      userPointer?.id
+        ? {
+            createdAt: Equal(userPointer.createdAt),
+            id: MoreThan(userPointer.id),
+          }
+        : 'TRUE',
+    )
+    .orderBy('u."createdAt"', 'ASC')
+    .addOrderBy('u."id"', 'ASC')
+    .take(limit);
+
+  const users = await qb.execute();
+
+  // nothing to do anymore, all the users are processed
+  if (users.length === 0) {
+    console.log('no users to process');
+    return null;
+  }
+
+  console.log(
+    'Queuing batch of users to process',
+    users.map((u) => u.id),
+  );
+  insertQueue.push(users);
+
+  const lastUser = users[users.length - 1];
+  return {
+    id: lastUser.id,
+    createdAt: lastUser.createdAt,
+  };
+};
+
+interface ViewWithDateTz extends View {
+  date: Date;
+  maxTimestamp: Date;
+}
+
+const FREEZE_DAYS_IN_A_WEEK = 2;
+const MISSED_LIMIT = 1;
+
+const computeReadingStreaksData = async (
+  con: DataSource,
+  user: Partial<User>,
+): Promise<DeepPartial<UserStreak>> => {
+  const timeZone = user.timezone ?? 'UTC';
+
+  const views: ViewWithDateTz[] = await con
+    .getRepository(View)
+    .createQueryBuilder()
+    .select([])
+    .addSelect('MAX(timestamp)', 'maxTimestamp')
+    .addSelect('DATE(timestamp::timestamptz AT TIME ZONE :timezone)', 'date')
+    .where({ userId: user.id, timestamp: Not(IsNull()) })
+    .setParameter('timezone', timeZone)
+    .addGroupBy('"date"')
+    .orderBy('"maxTimestamp"', 'ASC')
+    .getRawMany();
+
+  if (!views || views.length === 0) {
+    return {
+      currentStreak: 0,
+      totalStreak: 0,
+      maxStreak: 0,
+      lastViewAt: null,
+      userId: user.id,
+    };
+  }
+
+  const data = views.reduce(
+    (acc: DeepPartial<UserStreak>, item: ViewWithDateTz, index: number) => {
+      // prevent undefined results on first item
+      const difference =
+        index === 0 ? 1 : differenceInDays(item.date, acc.lastViewAt as Date);
+      const day = item.date.getDay();
+
+      // by default, increment streak
+      let currentStreak = acc.currentStreak + 1;
+
+      // restart streak if difference is greater than 1, taking freezes into account
+      if (
+        (day === Day.Sunday && difference > FREEZE_DAYS_IN_A_WEEK) ||
+        (day === Day.Monday &&
+          difference > FREEZE_DAYS_IN_A_WEEK + MISSED_LIMIT) ||
+        (day > Day.Monday && difference > MISSED_LIMIT)
+      ) {
+        currentStreak = 1;
+      }
```

Comment:
This logic is the same as the one in @sshanzel function (right?)
If it's the case can we uniform it